Add cluster reinstallation documentation (#238)
stolostron/siteconfig: docs/reinstallation/image_based_break_fix.md

## Overview

The Image-Based Break/Fix (IBBF) feature simplifies Single Node OpenShift (SNO) hardware replacement, minimizing downtime while preserving the cluster’s original identity. It retains critical cluster details, including identifiers, cryptographic keys (such as `kubeconfig`), and authentication credentials, allowing the replacement node to seamlessly assume the identity of the failed hardware.

Designed for like-for-like hardware replacements in SNO clusters installed using the Image-Based Install (IBI) method, IBBF introduces a GitOps-compatible, declarative API. Users can initiate hardware replacement with a single Git commit. Powered by the SiteConfig Operator and IBI Operator, this functionality enables cluster redeployment using the existing ClusterInstance custom resource (CR). 

With IBBF, OpenShift users gain a resilient, automated, and GitOps-native solution for quickly restoring SNO clusters after hardware failures.

## Requirements
To reinstall an OpenShift cluster using the SiteConfig Operator, ensure the following conditions are met:
- The cluster is a Single Node OpenShift (SNO) cluster installed using the Image-Based Install (IBI) provisioning method.

> - The faulty hardware is replaced with a new node that has identical specifications.

Reviewer: Should we call out that it should be installed with the same seed image as the original cluster?

Author: I think we can use a new seed image with a newer OpenShift verison. @carbonin can you keep me honest here?

2. **Update the `ClusterInstance` resource**: 
   - Set `spec.reinstall.preservationMode: ClusterIdentity`.
   - Assign a new, unique value to `spec.reinstall.generation`.
   - Update the `spec.nodes` object with the changed hardware information.
3. **Apply the changes** to initiate the reinstallation process.
4. **Monitor reinstallation progress**:
   - **Request validation**: The SiteConfig Operator validates the reinstallation request.
   - **Data preservation and cleanup**: The operator backs up labeled data.
   - **Cleanup**: The operator removes existing installation manifests. If this step fails within a specified time, deletion times out, and reinstallation stops.
   - **Data restoration**: The operator restores the preserved data. 
5. **Cluster reprovisioning**: The operator starts to re-provision the cluster by rendering new installation manifests from the referenced templates. Note that the [Image-Based Install Operator (IBIO)](https://github.com/openshift/image-based-install-operator) detects the preserved cluster identity data and uses it to generate the configuration ISO image.
6. **Cluster provisioning completes**: Confirm successful provisioning.

> 7. **Access the restored cluster**: Use the previous `kubeconfig` to access the reinstalled spoke cluster.

Author: Same response here.

> # Image-Based Break/Fix (IBBF) for SNO Hardware Replacement

Reviewer 2: Why have a whole separate doc for this?

Author: The flow for IBBF is slightly different with different checks - hence a separate, dedicated document to address it. Furthermore, it is also the primary requested feature from which the cluster reinstallation feature was built.

Author: @carbonin we decided to incorporate the IBBF doc under the same README.md. This avoids duplication of content.

     - **Request validation**: The SiteConfig operator validates the reinstallation request.  
     - **Data preservation**: The operator backs up labeled data.  
     - **Cleanup**: The operator removes existing installation manifests. If this step fails within a specified time, deletion times out, and reinstallation stops.  
     - **Data restoration**: The operator restores the preserved data.  
   - **Phase 2: Cluster provisioning** – Monitor the cluster provisioning processes through the `status.conditions` field in the `ClusterInstance` CR:  
     - **Manifest regeneration**: The SiteConfig operator generates new installation manifests from the referenced templates.  
     - **Cluster installation**: The operators consuming the installation manifests provision the cluster using the newly generated manifests. 
     :information_source: Note that the [Image-Based Install Operator (IBIO)](https://github.com/openshift/image-based-install-operator) detects the preserved cluster identity data and incorporates it into the configuration ISO image.  

> 5. **Completion verification**: Ensure the cluster is successfully provisioned and verify its availability. Use the previous `kubeconfig` to access the reinstalled spoke cluster.  

Reviewer 3: Or just "Verify changes". What does "verify its availability" mean?
```suggestion
5. **Complete verification**: Ensure the cluster is successfully provisioned and verify its availability. Use the previous `kubeconfig` to access the reinstalled spoke cluster.  
```

## Enabling Cluster Reinstallation in the SiteConfig Operator

To enable cluster reinstallation, update the `siteconfig-operator-configuration` `ConfigMap` resource by running the following command:

```sh
oc patch configmap siteconfig-operator-configuration \  
  -n $NAMESPACE \  
  --type=json \  
  -p '[{"op": "replace", "path": "/data/allowReinstalls", "value": "true"}]'
```

Verify that the `ConfigMap` resource is updated by running the following command:

```sh
oc get configmap siteconfig-operator-configuration -n $NAMESPACE -o yaml  
```

> ## Initiating the IBBF Cluster Reinstallation

Reviewer 3: ```suggestion
## Initiating the Image-Based Break/Fix Cluster Reinstallation
```

> To initiate the IBBF cluster reinstallation, update the `ClusterInstance` resource by setting the `spec.reinstall.generation` field and updating the `spec.nodes` object with the changed hardware information.

Reviewer 3: ```suggestion
To initiate the Image-Based Break/Fix cluster reinstallation, update the `ClusterInstance` resource by setting the `spec.reinstall.generation` field and updating the `spec.nodes` object with the changed hardware information.
```
